client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java

```diff
@@ -60,27 +84,66 @@ public void run() {
         }
         running = initConnectionWithServer();
         while (running) {
-                try {
-                    HttpRequest userRequest = HttpRequestSerializer.readFromInputStream(serverSocket.getInputStream());
-                    log.info("Client received new user request: {}", userRequest.toString());
-                } catch (IOException e) {
-                    log.error("Server is down");
-                    running = false;
-                }
+            processRequest();
         }
         stop();
     }
 
+    private void processRequest() {
+        try {
+            userRequest = HttpRequestSerializer.readFromInputStream(serverSocket.getInputStream());
+            log.info("Client received new user request: {}", userRequest.toString());
+        } catch (IOException e) {
+            log.error("Error reading user request from server", e);
+            running = false;
+        }
+        if (userRequest == null) {
+            return;
+        }
+
+        HttpResponse<String> serviceResponse = null;
+        try {
+            serviceSocket = new Socket("localhost", servicePort);
+            HttpClient httpClient = HttpClient.newHttpClient();
```

Comment:
Well, creating a new HTTP client is expensive operation so it's better to do that only once at startup.